src/tracker.rs

```diff
@@ -0,0 +1,300 @@
+        // 2. This iterator is then turned into a `RewindableTagsIter` which is a wrapper around the iterator
+        //    that allows us to rewind it to the beginning at our pleasure.
+        // 3. We use `dyn` to erase the type of the iterator, reducing binary size as it's better monomorphized.
+
+        {
+            // Get a slice of `&[impl AsRef<str>]`
+            let tags = tags.as_ref();
+            // Remember its length
+            let len = tags.len();
+            // Create an iterator that will infinitely cycle through the tags, returning `&str`s
+            let mut tags = tags.iter().map(|s| s.as_ref()).cycle();
+            // Create a stack allocated dyn iterator to put in the `RewindableTagsIter`
+            let tags: &mut dyn Iterator<Item = &str> = &mut tags;
+            // Construct a `RewindableTagsIter` that will allow us to rewind the iterator when needed
+            let tags = RewindableTagsIter { cursor: 0, tags, len };
+            // Now we can call `track`!
+            self.track(dd, metric, tags);
+        }
+
+        // Return the tags as-is for convenience
+        tags
+    }
+
+    fn track(&self, dd: &impl DogstatsdClient, metric: &str, mut tags: RewindableTagsIter) {
+        let mut lock = match self.state.as_ref() {
+            Some(state) => state.lock().unwrap(),
+            None => return,
+        };
+        let state = std::mem::replace(&mut *lock, TrackerState::Done);
+        match state {
+            TrackerState::Running {
+                mut seen,
+                mut cardinality_count,
+                actions,
+            } => {
+                let cardinality_grown = Self::update(&mut seen, metric, tags.rewind());
+                if cardinality_grown {
+                    cardinality_count += 1;
+                    if cardinality_count >= self.cardinality_threshold {
+                        drop(lock);
+                        Self::do_actions(dd, seen, actions, metric, tags.rewind());
+                        return;
+                    }
+                }
+                // Rebuild the running state if we didn't run the actions
+                *lock = TrackerState::Running {
+                    seen,
+                    cardinality_count,
+                    actions,
+                };
+            }
+            TrackerState::Done => {}
+        };
+    }
+
+    fn update(seen: &mut BTreeMap<String, Vec<BTreeSet<String>>>, metric: &str, tags: &mut RewindableTagsIter) -> bool {
+        let seen_tag_set = match seen.get_mut(metric) {
+            Some(seen_tag_set) => seen_tag_set,
+            None => {
+                seen.insert(
+                    metric.to_string(),
+                    vec![BTreeSet::from_iter(tags.rewind().map(|tag| tag.to_string()))],
+                );
+                return true;
+            }
+        };
+        let set_is_novel = seen_tag_set
```

Comment:
There was an useful comment on this line that has been removed in the last refactor, it helped me better understand what was going on:
```suggestion
        // Is this set of tags new for this metric?
        let set_is_novel = seen_tag_set
```